### ISSUE TITLE
Release memory after indexing is done?

### ISSUE THREAD

**Issue by alexislefebvre** [photoprism/photoprism#633]
When launching a PhotoPrism instance with Docker, it uses about 500 MB of memory.

Then when we start to index, we see this in logs: `classify: loading nasnet`.

And it looks like that the memory increase by about 1 GB.

That's fine. But it looks like this memory is never released again, even once indexing is done. Memory usage decreases only if we restart the container.

If nasnet use a significant quantity of memory, could PhotoPrism unload it automatically after some time?

**Comment by lastzero** [comment]
TensorFlow is managing this memory. Continuously loading and unloading it might cause a significant performance impact. Should be handled by TensorFlow and the OS ideally. Reports suggest this is what happens if you wait a little.

**Comment by lastzero** [comment]
You should find plenty of discussions on this in existing issues.

**Comment by alexislefebvre** [comment]
> TensorFlow is managing this memory. Continuously loading and unloading it might cause a significant performance impact.

That's why I suggested onloading after TensorFlow is idle for a finite amount of time.

> Should be handled by TensorFlow and the OS ideally. Reports suggest this is what happens if you wait a little.

What OS, host server or OS in Docker container? Do you know how much time should it take to release memory automatically?

> You should find plenty of discussions on this in existing issues.

Thanks, I searched and didn't found the same idea.

**Comment by lastzero** [comment]
> That's why I suggested onloading after TensorFlow is idle for a finite amount of time.

You could help us by figuring out if there is an API method for this. Their docs might also contain helpful information regarding memory usage and how to handle it.

Note that modern operating systems should automatically use paging to offload unused memory pages to disk, so that physical memory isn't blocked for other applications.

> What OS, host server or OS in Docker container? Do you know how much time should it take to release memory automatically?

We're running PhotoPrism on multiple servers in the cloud and at home and don't have memory issues in practice, so it wasn't a priority to investigate this in detail. Just because you see a high memory usage number in a task manager doesn't mean there is an actual issue, for example because TensorFlow or the operating system manage this transparently in the background.
 
> > You should find plenty of discussions on this in existing issues.
> 
> Thanks, I searched and didn't found the same idea.

Maybe it was discussed in our chat then or via email... probably related to memory issues on a Raspberry Pi. As an outcome, we've added system requirements to our docs (at least 4 GB physical RAM).

**Comment by lastzero** [comment]
For related issues see:
- https://github.com/photoprism/photoprism/issues/393#issuecomment-657502584
- https://github.com/photoprism/photoprism/issues/371
- https://github.com/photoprism/photoprism/issues/497

**Comment by alexislefebvre** [comment]
I didn't used PhotoPrism since yesterday, here is the memory usage:

```
# docker-compose logs --tail 3 photoprism
Attaching to photoprism_photoprism_1
photoprism_1             | time="2020-11-24T06:00:13Z" level=error msg="photo: record not found (estimate place)"
photoprism_1             | time="2020-11-24T06:00:13Z" level=error msg="photo: record not found (estimate place)"
photoprism_1             | time="2020-11-24T06:00:13Z" level=info msg="meta-worker: optimized 1 photos"
# docker-compose exec photoprism bash -c "ps -o pid,user,%mem,command ax | sort -b -k3 -r"
    6 root     18.5 photoprism start
  847 root      0.0 sort -b -k3 -r
  846 root      0.0 ps -o pid,user,%mem,command ax
    1 root      0.0 bash /entrypoint.sh photoprism start
  840 root      0.0 bash -c ps -o pid,user,%mem,command ax | sort -b -k3 -r
  PID USER     %MEM COMMAND
```

So after 6 hours without any operation, it still uses 18.5 % of my memory.

---

So I restarted the container:

```
# cd /var/photoprism/ ; docker-compose restart photoprism
Restarting photoprism_photoprism_1 ... done
# docker-compose logs --tail 5 photoprism
Attaching to photoprism_photoprism_1
photoprism_1             | time="2020-11-24T12:30:43Z" level=info msg="shutting down web server"
photoprism_1             | time="2020-11-24T12:30:50Z" level=info msg="read-only mode enabled"
photoprism_1             | time="2020-11-24T12:30:50Z" level=info msg="webdav: /originals/ waiting for connection"
photoprism_1             | time="2020-11-24T12:30:50Z" level=info msg="webdav: /import/ not available in read-only mode"
photoprism_1             | time="2020-11-24T12:30:50Z" level=info msg="starting web server at 0.0.0.0:2342"
# docker-compose exec photoprism bash -c "ps -o pid,user,%mem,command ax | sort -b -k3 -r"
    9 root      5.1 photoprism start
   25 root      0.0 sort -b -k3 -r
   24 root      0.0 ps -o pid,user,%mem,command ax
    1 root      0.0 bash /entrypoint.sh photoprism start
   18 root      0.0 bash -c ps -o pid,user,%mem,command ax | sort -b -k3 -r
  PID USER     %MEM COMMAND
```

And the memory usage dropped.

**Comment by lastzero** [comment]
See Release Notes of Go 1.12:  https://go.dev/doc/go1.12

> On Linux, the runtime now uses `MADV_FREE` to release unused memory. This is more efficient but may result in higher reported RSS. The kernel will reclaim the unused data when it is needed. To revert to the Go 1.11 behavior (`MADV_DONTNEED`), set the environment variable `GODEBUG=madvdontneed=1`.

Maybe that is the "issue"? Please verify as we will otherwise close this, assuming that it is only a problem with how the memory consumption is reported.

**Comment by alexislefebvre** [comment]
@lastzero what are we supposed to check? Go 1.12 was [released](https://go.dev/doc/devel/release#go1.12) one year before I opened this issue, so this behaviour was already present. Does the system RAM must be full to trigger the unloading of nasnet and release the memory?

**Comment by lastzero** [comment]
> @lastzero what are we supposed to check? Go 1.12 was [released](https://go.dev/doc/devel/release#go1.12) one year before I opened this issue, so this behaviour was already present.

Exactly. You reported that memory is not released, and the Go docs say it will look that way, and yet they implemented it like that simply because it improves performance.

> Does the system RAM must be full to trigger the unloading of nasnet and release the memory?

The Linux kernel is responsible for freeing memory when needed, which I understand can be several hours later. If you have other information, please share it. 👍

**Comment by alexislefebvre** [comment]
I started Photoprism and indexing yesterday at 18:00, then today at noon I used [`tail /dev/zero`](https://unix.stackexchange.com/questions/99334/how-to-fill-90-of-the-free-memory/254976#254976) in order to fill the RAM (note the red spike of the swap at noon), the memory usage of `apps` as reported by munin was unchanged, the memory used by `apps` lowered only when I shut down the Photoprism container (on the far right of the graph).

![image](https://user-images.githubusercontent.com/2071331/177982210-76599a8e-c819-41e4-b4cf-7c8829c914fb.png)

So it looks like that filling the RAM didn't trigger a release of memory by Go.

Note that Go changed again the memory management in [1.16](https://go.dev/doc/go1.16#runtime):

>  On Linux, the runtime now defaults to releasing memory to the operating system promptly (using `MADV_DONTNEED`), rather than lazily when the operating system is under memory pressure (using `MADV_FREE`). This means process-level memory statistics like RSS will more accurately reflect the amount of physical memory being used by Go processes. Systems that are currently using `GODEBUG=madvdontneed=1` to improve memory monitoring behavior no longer need to set this environment variable.

**Comment by lastzero** [comment]
Interesting, thanks for testing! I'm still not sure if it makes sense to "change" this (or what to change in the first place):

- unloading and reloading the TensorFlow model, for example, is an expensive operation
- this is especially bad for users who have background indexing set up that might run every 5 minutes

**Comment by ninbura** [comment]
Just chiming in on this; seeing the same thing. After indexing 6000 images my memory rose to 50GB used and remained there for 24 hours even after the indexing subsided. I'm guessing I'll have to build in a process that takes down all my containers on a semi-regular basis to release the memory, and then bring everything back up.

Not ideal, could interrupt large indexes. Would be nice if there was a process built-in that could recognize that nothing is being currently indexed and clear the memory. I'm sure it sounds much more simple than it is.

**Comment by lastzero** [comment]
@ninbura We have large libraries and at no point in time 50 GB of RAM were used. Do you have large (panoramic) images or how could this happen? If you see something like this, it may be a virtual memory reservation that has no effect on physical memory usage. Do you have practical experience with Linux/Unix memory management?

**Comment by ninbura** [comment]
I'm running my instance via docker on wsl. As the index ran the memory climbed to 50gb  and stayed there until I ran `wsl --shutdown`.

I used HWiNFO to monitor memory usage. You're supposing that the container just had that much memory reserved but it wasn't being used?

**Comment by lastzero** [comment]
@ninbura Note that WSL has a dynamic memory allocation and I would guess that the maximum usable memory is not reduced once the Linux VM running on Windows is given more memory. However, it can be shared with Windows so that the same memory can be used by other processes. This is possible due to the way modern CPUs work. Unfortunately, we are busy with other tasks, so I don't have time to do any research on WSL memory management. In general, when you post such statistics, it would be very helpful if you also share how much total memory you have, if swap is used, and what CPU you are using. It would also be good to know if you have any files - such as extremely large images - that would require a lot of memory during indexing (but not after). I'm pretty sure you don't need a process that shuts down Docker containers and then starts them up again. That would usually take up additional resources rather than reduce usage.

**Comment by ninbura** [comment]
Apologies for the lack of information. I had just assumed I was running into the same issue as OP.

motherboard - asrock x399m 
cpu - amd threadripper 2970wx
memory - 128gb ddr4 @ 3200Mhz

Grand majority of images are 12mp, encoded in heif, and converted to jpeg @ quality 85 to be viewed in web.

Looks like the container has access to half my system memory at 64GB with 16GB of swap:
![image](https://github.com/photoprism/photoprism/assets/58058942/c62faaa1-deed-42fe-8a82-ddf527789cfb)

I didn't think to take a look inside the container before taking it down & just took hwinfo at face value. I'll delete my storage and sidecar files, re-index the library, and examine memory usage directly in the container.

**Comment by lastzero** [comment]
With the threadripper, as the name suggests, you will have a lot of indexing workers by default as that depends on the number of cores available. Each of those workers requires memory and so Windows will make it available to the VM. If your goal is to reduce peak memory usage, then limit the number of workers using the related config option (see our docs and troubleshooting guide for details).

**Comment by MyWebmasterD** [comment]
Hi guys,
I have the same issue here..
Photoprism on qnap ts-473a with docker, 6GB of memory in constant usage after indexing.

Anyone understood what is causing the issue?

**Comment by alexislefebvre** [comment]
> Hi guys, I have the same issue here.. Photoprism on qnap ts-473a with docker, 6GB of memory in constant usage after indexing.
> 
> Anyone understood what is causing the issue?

To my knowledge, no. Several users observed that TensorFlow stay in memory after the indexing is finished. That's the only thing we are sure of.

Is it an issue from TensorFlow? Would it be possible to unload TensorFlow from the memory? Or, to mitigate the issue, *mark* the memory used by TF has low priority so that if another process needs memory, TF would be unloaded from memory instead of being moved to the swap space? (this last sentence is pure speculation)

I think that ideally, PhotoPrism should have an opt-in option like *unload TF from memory when the indexer has been inactive for X hours*. But I don't know if that's possible. If someone knows how to do it, a PR would be welcome.

**Comment by lastzero** [comment]
From what others reported, TensorFlow might automatically free the momory after some hours (see existing GitHub issues and discussions). But even if it does not, the kernel should write the unused model to disk so the physical memory can be used by your other applications – which is why our documentation recommends to configure at least 4 GB of swap. The model can be quickly restored from there if needed, generally faster than loading it from the original saved model file. Likewise, it can be that some memory is used by the kernel / filesystem for caching files in order to improve performance. This also shouldn't be a problem if other applications need to use the physical memory at a later time.
